app/jobs/autoscaling.py

```diff
@@ -198,9 +208,146 @@ def set_desired_capacity(asg, compute_desired_instances, can_scale=True):
     if can_scale:
         if DEBUG:
             print cmd
+        if num_desired < previous_desired:
+            start_draining(asg, previous_desired - num_desired)
+        elif num_desired > previous_desired:
+            stop_draining(asg, num_desired - previous_desired)
         aws_command(cmd)
 
 
+def instances_in(asg):
+    return [item["InstanceId"] for item in asg["Instances"]]
+
+
+def unixtime_now():
+    return int(time.time())
+
+
+def iso2unixtime(iso_str):
+    return int(dateutil.parser.parse(iso_str).strftime('%s'))
+
+
+def add_draining_tag(instance_ids):
+    cmd = "aws ec2 create-tags --resources {list_instances} --tags Key={draining_tag},Value={timestamp}"
+    cmd = cmd.format(list_instances=' '.join(instance_ids), draining_tag=DRAINING_TAG, timestamp=unixtime_now())
+    aws_command(cmd)
+
+
+def remove_draining_tag(instance_ids):
+    delete_tags_from_instances(instance_ids, [DRAINING_TAG])
+
+
+def start_draining(asg, num_instances):
+    instance_ids = instances_to_drain(asg, num_instances)
+    add_draining_tag(instance_ids)
+
+
+def stop_draining(asg, num_instances):
+    instance_ids = instances_to_rescue(asg, num_instances)
+    if instance_ids:
+        print "Stopping drainage of the following instances: " + ",".join(instance_ids)
+        add_termination_protection(instance_ids, asg)
+        # Note: if autoscaling.py ran too frequently, there would be a risk that an instance enters Terminating state between the time
+        # of the health check in instances_to_rescue and the time of add_termination_protection, in which case add_termination_protection
+        # would throw an error. But because autoscaling.py runs only infrequently, we can be sure that there is no scaling decision in limbo
+        # from the last execution of autoscaling.py.
+        remove_draining_tag(instance_ids)
+
+
+def instances_to_drain(asg, num_instances):
+    instance_ids = instances_in(asg)
+    cmd = "aws ec2 describe-instances --instance-ids {list_instances} --query 'Reservations[*].Instances[*].[InstanceId,LaunchTime]' --no-paginate"
+    cmd = cmd.format(list_instances=' '.join(instance_ids))
+    aws_response = json.loads(aws_command(cmd))
+    zipped_instance_ids_and_launch_times = [(instance[0], iso2unixtime(instance[1])) for nest in aws_response for instance in nest]
+    instance_ids_sorted_by_increasing_launch_time = [item[0] for item in sorted(zipped_instance_ids_and_launch_times, key=itemgetter(1))]
+    return instance_ids_sorted_by_increasing_launch_time[:num_instances]
+
+
+def instances_to_rescue(asg, num_instances):
+    '''
+    Rescue the instances that have been draining for the least amount of time.
+    That way we can probably terminate the non-rescued instances sooner.
+    '''
+    draining_instances = get_draining_servers(asg)
+    healthy_instance_ids = [inst["InstanceId"] for inst in asg["Instances"] if inst["HealthStatus"] == "Healthy" and inst["LifecycleState"] != "Terminating"]
```

Comment:
So here I think rather than collecting instances based on LifecycleState, you want instead to collect instances that still have their scale-in protection set.

Reply:
In fact I would change get_draining instances to filter to those.   Once protection has been removed, the instance is no longer draining.  It is drained.

Reply:
makes sense, thanks